bibtex-actions.el

```diff
@@ -531,35 +531,86 @@ With prefix, rebuild the cache before offering candidates."
+
+(defun bibtex-actions--add-local-watches (func &optional extra-local-files)
   "Add watches for the files that contribute to the local cache
-The callback FUNC is run when a chage in one of the local bibliography files is notified."
-       (let ((buffer (buffer-name)))
-         (seq-map (lambda (bibfile)
-                    (file-notify-add-watch bibfile '(change)
-                                           (lambda (x) (unless (eq 'stopped (cadr x))
-                                                    (with-current-buffer buffer (funcall func))))))
-                  (bibtex-actions--local-files-to-cache))))
+The callback FUNC is run when a chage in one of the local bibliography files is notified.
+The optional argument EXTRA-LOCAL-FILES is a list of paths. If passed watches will also be
+placed on these files."
+  (let ((buffer (buffer-name)))
+    (setq bibtex-actions--local-watches
+          (seq-map (lambda (bibfile)
+                     (file-notify-add-watch bibfile '(change)
+                                            (lambda (x) (unless (eq 'stopped (cadr x))
+                                                     (with-current-buffer buffer (funcall func))))))
+                   (seq-concatenate 'list (bibtex-actions--local-files-to-cache) extra-local-files)))))
 
 ;;;###autoload
-(defun bibtex-actions-with-filenotify-local (func)
+(defun bibtex-actions-with-filenotify-local (func &optional extra-local-files)
   "Hook to add watches on buffer local bib files and remove them when the buffer is killed
-The callback FUNC is run when a change is notified"
-  (let ((descs (bibtex-actions--add-local-watches func)))
+The callback FUNC is run when a change is notified.
+The optional argument EXTRA-LOCAL-FILES is a list of paths. If passed watches will also be
+placed on these files.
+The watches are added only if`bibtex-actions--local-watches'
+has the default value `uninitialized'. This is to ensure that duplicate watches aren't added. This means
+a mode hook containing this function can run several times without adding duplicate watches."
+  (when (eq 'uninitialized bibtex-actions--local-watches)
+    (bibtex-actions--add-local-watches func extra-local-files)
     (add-hook 'kill-buffer-hook
-              (lambda () (seq-map #'file-notify-rm-watch descs)) nil t)))
+              (lambda () (seq-map #'file-notify-rm-watch bibtex-actions--local-watches)
+                (setq bibtex-actions--local-watches 'uninitialized))
+              nil t)))
+
+(defun bibtex-actions--with-filenotify-local-refresh (func &optional extra-local-files)
+  "Refresh the watches on the local bib files.
+The callback FUNC is run when a change is notified.
+The optional argument EXTRA-LOCAL-FILES is a list of paths. If passed watches will also be
+placed on these files.
+This function only needs to be called if a local bib file has been added or removed."
+  (seq-map #'file-notify-rm-watch bibtex-actions--local-watches)
+  (reftex-access-scan-info t)
```

Comment:
This is giving me a warning.

Reply:
From my going through the `bibtex-completion` code, `bibtex-completion-find-local-bibliography` use `reftex` to find the local bibliography. There isn't a separate both so I assume it is used for `org` too.  `(reftex-access-scan-info 1)` tells `reftex` to rescan so addition or removal of a bib file is detected. Without it there `bibtex-completion-find-local-bibliography` and our functions using it will give the old files.

I didn't get the warning though.

Reply:
I'm using Emacs 28 nativecomp, and as I said, got a compilation failure.

But declaring the function fixes that.

Reply:
I made this change.